### PR TITLE
Intercept cloud

### PR DESCRIPTION
**Pull request by ArkamJ**
_Issue #, if available:_

_Description of changes:_

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/src/sync/ComplexObjUtils.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+			if (typeof obj[key] === 'object') {
+				if (!isEmpty(obj[key])) {
+					iterate(obj[key]);
+				}
+			}
+		});
+	};
+	await iterate(cloudObject);
+	return fileList;
+}
```
Comment (iartemiev) [inline packages/datastore/src/sync/ComplexObjUtils.ts:82]: I haven't tested this, but something along these lines should work:

```typescript
async function downloadComplexObjects(
	cloudObject
): Promise<Array<ComplexObject>> {
	console.log('Entered downloadComplexObjects');
	const keys = [];

	const iterate = async obj => {
		Object.keys(obj).forEach(async key => {
			const isJsonString = tryParseJSON(obj[key]);
			if (isJsonString) {
				if (isJsonString.s3Key) {
					const s3Key = isJsonString.s3Key;
					keys.push(s3Key);
				}
			}

			if (typeof obj[key] === 'object') {
				if (!isEmpty(obj[key])) {
					iterate(obj[key]);
				}
			}
		});
	};
	await iterate(cloudObject);

	const fileList = await Promise.all(
		keys.map(async key => {
			const result = await storageCategory.get(key, {
				download: true,
			});
			const file = result['Body'];
			return { file, s3Key: key };
		})
	);

	return fileList;
}
```

Comment (ArkamJ) [reply packages/datastore/src/sync/ComplexObjUtils.ts:82]: That worked! Thank you!

File: packages/datastore/src/sync/ComplexObjUtils.ts [hunk 2]
```diff
@@ -18,6 +20,34 @@ function blobToFile(blob: Blob, fileName: string): File {
...
+	const fileList = await Promise.all(
+		keys.map(async key => {
+			const buf = await fileToArrayBuffer(blob);
+			var wordArray = CryptoJS.lib.WordArray.create(buf);
+			const out = md5(wordArray).toString();
+			return out;
+		})
+	);
+	return fileList[0];
+}
```
Comment (iartemiev) [inline packages/datastore/src/sync/ComplexObjUtils.ts:49]: if it's just the one blob `Promise.all` and `.map` are redundant.

You should be able to just do:
```typescript
async function getEtag(blob) {
  const buf = await fileToArrayBuffer(blob);
  const wordArray = CryptoJS.lib.WordArray.create(buf);
  const out = md5(wordArray).toString();
  return out;
}
```

File: packages/datastore/src/sync/ComplexObjUtils.ts [hunk 3]
```diff
@@ -113,11 +152,19 @@ async function downloadComplexObjects(
...
+			const complexObject = await storageCategory
+				.get(key, {
+					download: true,
+				})
+				.then(result => {
+					const file = blobToFile(result['Body'], key.split('/')[2]);
+					return { file, s3Key: key };
+				})
+				.catch(err => {
+					// If 404 error, file not in S3 Bucket, return nothing as the item is of type DELETE
```
Comment (iartemiev) [inline packages/datastore/src/sync/ComplexObjUtils.ts:164]: If possible, let's check that it's a 404 specifically.
If it is - `return`
For other errors `throw err`

File: packages/datastore/src/sync/ComplexObjUtils.ts (hunk 2 above)
Comment (ArkamJ) [reply packages/datastore/src/sync/ComplexObjUtils.ts]: Ohh got it. Thank you!

**Pull request merged**